### PR TITLE
Add set_equivalence_ratio to SolutionArray objects

### LINKED ISSUE #604: Add `set_equivalence_ratio` to `SolutionArray` objects
Currently, the Python `SolutionArray` class has no equivalent of `ThermoPhase.set_equivalence_ratio`. A version of this function could be added to the `SolutionArray` class so that one can set the equivalence ratio to a range of values. This would make calculation of an adiabatic flame temperature curve very simple:
```py
states = ct.SolutionArray(gas, 100)
phi = np.arange(0, 2, 100)
states.set_equivalence_ratio(phi, 'CH4:1.0', 'O2:1.0, N2:3.76')
states.equilibrate('HP')
plt.plot(phi, states.T)
```

This function could also enable specifying a range of fuel and/or oxidizer compositions, in the case where a `dict` is used for the composition:
```py
states.set_equivalence_ratio(phi, {'CH4': np.linspace(0, 1, 100), 'H2': np.linspace(1, 0, 100)})
```

### PR DESCRIPTION
**Pull request by ischoegl**
Please fill in the issue number this pull request is fixing:

Fixes #604

Changes proposed in this pull request:
- Wrap `ThermoPhase.set_equivalence_ratio`, and check dimensions
- Add unit test

### CODE REVIEW

File: interfaces/cython/cantera/composite.py [hunk 1]
```diff
@@ -565,6 +566,30 @@ def equilibrate(self, *args, **kwargs):
...
+
+        Note that *phi* either needs to be a scalar value or dimensions have
+        to be matched to the SolutionArray.
+        """
+
+        # test arguments
+        if isinstance(phi, _numbers.Real):
+            phi = phi * np.ones(self._shape)
+        elif isinstance(phi, np.ndarray):
+            if phi.shape!=self._shape:
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]:
```suggestion
            if phi.shape != self._shape:
```

File: interfaces/cython/cantera/composite.py [hunk 2]
```diff
@@ -565,6 +566,30 @@ def equilibrate(self, *args, **kwargs):
...
+                raise ValueError('Shape mismatch.')
+        else:
+            if len(phi)!=len(self._indices):
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]:
```suggestion
            if len(phi) != len(self._indices):
```

File: interfaces/cython/cantera/composite.py [hunk 3]
```diff
@@ -565,6 +566,30 @@ def equilibrate(self, *args, **kwargs):
...
+                raise ValueError('Incompatible dimensions.')
```
Comment (speth) [inline interfaces/cython/cantera/composite.py]: The alternative to this block is to use the same approach as is used in the state setting functions:
```python
    phi, _ = np.broadcast_arrays(phi, self._output_dummy)
```
This also allows things like broadcasting a `phi` array to a higher-dimensional `SolutionArray`, e.g. from 1xN to MxN.

**Review by speth — CHANGES_REQUESTED** [review]
I had an alternative suggestion for mapping the input equivalence ratio to an array of the appropriate shape.

File: interfaces/cython/cantera/composite.py (hunk 3 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: NP. I was looking for something existing but obviously missed that one ...

File: interfaces/cython/cantera/composite.py (hunk 2 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: N/A.

File: interfaces/cython/cantera/composite.py (hunk 1 above)
Comment (ischoegl) [reply interfaces/cython/cantera/composite.py]: N/A.

**Comment by ischoegl** [comment]
Done and pushed. Let's see whether unit tests do their job ...

**Comment by ischoegl** [comment]
unit tests work ;)

**Review by speth — APPROVED** [review]

**Pull request merged**